examples/pytorch/rgcn-hetero/model.py

```diff
@@ -0,0 +1,199 @@
+
+class RelGraphConvLayer(nn.Module):
+    r"""Relational graph convolution layer.
+
+    Parameters
+    ----------
+    in_feat : int
+        Input feature size.
+    out_feat : int
+        Output feature size.
+    rel_names : list[str]
+        Relation names.
+    num_bases : int, optional
+        Number of bases. If is none, use number of relations. Default: None.
+    weight : bool, optional
+        True if a linear layer is applied after message passing. Default: True
+    bias : bool, optional
+        True if bias is added. Default: True
+    activation : callable, optional
+        Activation function. Default: None
+    self_loop : bool, optional
+        True to include self loop message. Default: False
+    dropout : float, optional
+        Dropout rate. Default: 0.0
+    """
+    def __init__(self,
+                 in_feat,
+                 out_feat,
+                 rel_names,
+                 num_bases,
+                 *,
+                 weight=True,
+                 bias=True,
+                 activation=None,
+                 self_loop=False,
+                 dropout=0.0):
+        super(RelGraphConvLayer, self).__init__()
+        self.in_feat = in_feat
+        self.out_feat = out_feat
+        self.rel_names = rel_names
+        self.num_bases = num_bases
+        self.bias = bias
+        self.activation = activation
+        self.self_loop = self_loop
+
+        self.conv = dglnn.HeteroGraphConv({
+                rel : dglnn.GraphConv(in_feat, out_feat, norm='right', weight=False, bias=False)
+                for rel in rel_names
+            })
+
+        if weight:
+            self.basis = dglnn.WeightBasis(in_feat, out_feat, num_bases, len(self.rel_names))
+        else:
+            self.basis = None
+
+        # bias
+        if bias:
+            self.h_bias = nn.Parameter(th.Tensor(out_feat))
+            nn.init.zeros_(self.h_bias)
+
+        # weight for self loop
+        if self.self_loop:
+            self.loop_weight = nn.Parameter(th.Tensor(in_feat, out_feat))
+            nn.init.xavier_uniform_(self.loop_weight,
+                                    gain=nn.init.calculate_gain('relu'))
+
+        self.dropout = nn.Dropout(dropout)
+
+    def forward(self, g, inputs):
+        """Forward computation
+
+        Parameters
+        ----------
+        g : DGLHeteroGraph
+            Input graph.
+        inputs : dict[str, torch.Tensor]
+            Node feature for each node type.
+
+        Returns
+        -------
+        dict[str, torch.Tensor]
+            New node features for each node type.
+        """
+        g = g.local_var()
+        if self.basis:
+            weights = self.basis()
+            assert weights.shape == (len(self.rel_names), self.in_feat, self.out_feat)
+            wdict = {self.rel_names[i] : w.squeeze(0) for i, w in enumerate(th.split(weights, 1, dim=0))}
+            for rel, mod in self.conv.mods.items():
+                mod.weight = wdict[rel]
```

Comment:
It is basically a form of weight sharing. I googled a little bit about what's the best practice. From the answer of the PyTorch developer (https://discuss.pytorch.org/t/how-to-create-model-with-sharing-weight/398), he suggested defining models like this:
```python
class MyModel(nn.Module):
    def __init__(self):
        self.base = ...
        self.head_A = ...
        self.head_B = ...

    def forward(self, input1, input2):
        return self.head_A(self.base(input1)), self.head_B(self.base(input2))
```
In our case, the `self.basis` instance is the `self.base` with no inputs. If adopting the above approach, I need to add an extra argument `weight` to `nn.GraphConv` for externally provided weight tensor. This complicates the `HeteroGraphConv` module since each submodule will have some extra arguments.

My proposal to this is extending the `HeteroGraphConv.forward` to allow more complicated argument specification. For example, we could have:
```python
hetconv = dglnn.HeteroGraphConv({'plays' : mod1, 'sells' : mod2, 'follows' : mod3})
node_inputs = {'user' : ..., 'game' : ...}
mod_args = {'plays' : (a1, a2), 'sells' : (b1, b2), 'follows' : (c1, c2)}
```
And the `mod_args` will be dispatched to each modules upon invocation. In this way, we can resolve your concern like the following:
```python
hetconv = dglnn.HeteroGraphConv({'plays' : mod1, 'sells' : mod2, 'follows' : mod3})
node_inputs = {'user' : ..., 'game' : ...}
h = hetconv(node_inputs, mod_args=wdict)
```

Thoughts?

Reply:
> It is basically a form of weight sharing. I googled a little bit about what's the best practice. From the answer of the PyTorch developer (https://discuss.pytorch.org/t/how-to-create-model-with-sharing-weight/398), he suggested defining models like this:
> 
> ```python
> class MyModel(nn.Module):
>     def __init__(self):
>         self.base = ...
>         self.head_A = ...
>         self.head_B = ...
> 
>     def forward(self, input1, input2):
>         return self.head_A(self.base(input1)), self.head_B(self.base(input2))
> ```
> 
> In our case, the `self.basis` instance is the `self.base` with no inputs. If adopting the above approach, I need to add an extra argument `weight` to `nn.GraphConv` for externally provided weight tensor. This complicates the `HeteroGraphConv` module since each submodule will have some extra arguments.
> 
> My proposal to this is extending the `HeteroGraphConv.forward` to allow more complicated argument specification. For example, we could have:
> 
> ```python
> hetconv = dglnn.HeteroGraphConv({'plays' : mod1, 'sells' : mod2, 'follows' : mod3})
> node_inputs = {'user' : ..., 'game' : ...}
> mod_args = {'plays' : (a1, a2), 'sells' : (b1, b2), 'follows' : (c1, c2)}
> ```
> 
> And the `mod_args` will be dispatched to each modules upon invocation. In this way, we can resolve your concern like the following:
> 
> ```python
> hetconv = dglnn.HeteroGraphConv({'plays' : mod1, 'sells' : mod2, 'follows' : mod3})
> node_inputs = {'user' : ..., 'game' : ...}
> h = hetconv(node_inputs, mod_args=wdict)

This one looks better. I like this one.

> ```
> 
> Thoughts?